lib/core/src/Cardano/Pool/DB.hs

```diff
@@ -166,6 +188,87 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
         -- For a Sqlite DB, this would be "run a query inside a transaction".
     }
 
+-- | Given the /latest/ registration and retirement certificates for a pool,
+--   determine the pool's current registration status based on the relative
+--   order in which the certificates were published.
+--
+-- If two certificates are supplied, then:
+--
+--   * the certificates must be from the same pool.
+--   * the publication times must be non-equal.
+--
+-- This function determines order of precedence according to the "pool
+-- inference rule", as described in "A Formal Specification of the Cardano
+-- Ledger":
+--
+-- https://hydra.iohk.io/build/3202141/download/1/ledger-spec.pdf
+--
+determinePoolRegistrationStatus
+    :: Ord certificatePublicationTime
+    => Maybe (certificatePublicationTime, PoolRegistrationCertificate)
+    -> Maybe (certificatePublicationTime, PoolRetirementCertificate)
+    -> PoolRegistrationStatus
+determinePoolRegistrationStatus mReg mRet = case (mReg, mRet) of
+    (Nothing, _) ->
+        PoolNotRegistered
+    (Just (_, regCert), Nothing) ->
+        PoolRegistered regCert
+    (Just (regTime, regCert), Just (retTime, retCert))
+        | regPoolId /= retPoolId ->
+            -- Comparing certificates from different pools is a programming
+            -- error.
+            differentPoolsError
```

Comment:
You may as well remove the comment and inline the differentPoolsError string -- it's self-documenting.

Reply:
> You may as well remove the comment and inline the differentPoolsError string -- it's self-documenting.

Thanks! I've removed the comment, and revised the error string to make it clear that it's a programming error.

(The only thing I haven't changed is to inline `differentPoolsError`, in order to keep the pattern matching fairly compact.)